src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -451,35 +473,111 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
             // We start by binding everything that is not obviously reorderable as a temporary, and
             // then run an optimizer to remove unnecessary temporaries.
 
-            BoundExpression[] actualArguments = new BoundExpression[parameters.Length]; // The actual arguments that will be passed; one actual argument per formal parameter.
-            ArrayBuilder<BoundAssignmentOperator> storesToTemps = ArrayBuilder<BoundAssignmentOperator>.GetInstance(rewrittenArguments.Length);
-            ArrayBuilder<RefKind> refKinds = ArrayBuilder<RefKind>.GetInstance(parameters.Length, RefKind.None);
+            // Interpolated string handler conversions introduce another wrinkle to this algorithm, as
+            // they can reuse the receiver or parameters from the call to perform the conversion
+            // as part of the argument list. Our saving grace here is that these conversions _must_
+            // come lexically after the arguments they use, so we don't have to deal with the param
+            // arrays or optional parameters before lowering the conversions. However, because there can
+            // be things such as lambda conversions that are not safe to evaluate twice, we have to skip
+            // the temp optimization steps to ensure that we don't do these double evaluations.
+
+            var actualArguments = new BoundExpression[parameters.Length]; // The actual arguments that will be passed; one actual argument per formal parameter.
+            var storesToTemps = ArrayBuilder<BoundExpression>.GetInstance(arguments.Length);
+            var refKinds = ArrayBuilder<RefKind>.GetInstance(parameters.Length, RefKind.None);
+            var requiresInstanceReceiver = methodOrIndexer.RequiresInstanceReceiver() && methodOrIndexer is not MethodSymbol { MethodKind: MethodKind.Constructor };
 
             // Step one: Store everything that is non-trivial into a temporary; record the
             // stores in storesToTemps and make the actual argument a reference to the temp.
             // Do not yet attempt to deal with params arrays or optional arguments.
-            BuildStoresToTemps(
-                expanded,
-                argsToParamsOpt,
-                parameters,
-                argumentRefKindsOpt,
-                rewrittenArguments,
-                forceLambdaSpilling: false, // lambda conversions can be re-orderd in calls without side affects
-                actualArguments,
-                refKinds,
-                storesToTemps);
-
+            if (interpolatedStringHandlerArgumentIndex < 0)
+            {
+                // No handler conversions, we can just store all directly to temps
+                VisitAndStoreToTemps(
```

Comment:
>VisitAndStoreToTemps [](http://example.com/codeflow?start=16&length=20)

If I understand correctly, we are conflating creation of temps needed to reorder named arguments with temps needed to create handlers. I think it would be simple to separate these two concerns. First, we do rewrite for handlers and patch the arguments, then let existing logic to kick in to handle reordering and let it deal with its own temp stores etc.